Upload of a resource's data file (creation and update)

bobeal opened the pull request:
Hello,

As discussed two weeks ago, here is the PR for the management of a resource's data file.

Do not hesitate to contact me if some things have to be re-worked.

Regards,

Benoit.

bobeal committed:
Upload of a resource's data file (creation and update)

DavidLeoni commented:
Thank you very much, looks like a clean commit! 

I have one question about file sizes: Have you tried to upload files of several megabytes? I think a test case for that would make us sure everything is going to work properly with ckan (In case try to generate the file on the fly to prevent storing too big stuff in the repo,  as far as I remember datahub.io supports max 100mb).

DavidLeoni commented on src/main/java/eu/trentorise/opendata/jackan/model/CkanResourceBase.java:
```diff
@@ -383,6 +395,33 @@ public void setUrl(String url) {
+            this.mimetype = mediaType.getBaseType().toString();
+            this.format = mediaType.getSubtype();
+        } catch (FileNotFoundException e) {
+            LOG.log(Level.WARNING, "Unable to load file {0}", upload.getName());
```
I think it would be better to rethow in a JackanException

bobeal replied:
I found it a little harsh for the user (as it's only setting mime type and format that failed) but ok for it (and, yes, it's the only way to tell it something gone wrong).

DavidLeoni replied:
Fact is I don't know how ckan behaves when uploading files. Questions that came up in my mind: Is mimetype required? Is format required? Does it have to actually match an uploaded resource format? Should format be only in catalog's allowed formats? Depending on the answer you may keep mimetype optional with just the log or force somewhat the user to include mimetype and format. In general when developing Jackan I adopted minimal automations because ckan already has it's own fair amount of quirks which I didn't want to complicate further . Of course mimetype detection is useful, so it is only a matter of how to implement it so that if issues occur it's clear where the problem is.

bobeal replied:
From what I've tested : 
- Mime type is not required and just seems to be informative
- Format is not required but is used to set the type of the resource file. If it is not set, the resource is shown as « Data » and, for instance, default views are not created. 
- CKAN is not restrictive about the Format field and just displays what we passed to him.

Maybe a good compromise would be to let this mimetype and format guessing up to the user (eg by adding a boolean parameter to the setUpload method) ? And if it fails, we throw an exception and the user will do what seems appropriate to him.

DavidLeoni replied:
> Maybe a good compromise would be to let this mimetype and format guessing up to the user (eg by adding a boolean parameter to the setUpload method) ? And if it fails, we throw an exception and the user will do what seems appropriate to him.

That sounds perfect!

DavidLeoni commented on src/main/java/eu/trentorise/opendata/jackan/model/CkanResourceBase.java:
```diff
@@ -383,6 +395,33 @@ public void setUrl(String url) {
+        } catch (FileNotFoundException e) {
+            LOG.log(Level.WARNING, "Unable to load file {0}", upload.getName());
+        } catch (IOException e) {
+            LOG.log(Level.WARNING, "Unable to detect mime type and format for file " + upload.getName() + " : {1}", e);
```
I think it would be better to retrhow in a JackanException

DavidLeoni commented on src/main/java/eu/trentorise/opendata/jackan/model/CkanResourceBase.java:
```diff
@@ -383,6 +395,33 @@ public void setUrl(String url) {
+     */
+    public void setUpload(File upload) {
+        this.upload = upload;
+        this.size = String.valueOf(upload.length());
```
Can upload be null? if so here it shouldn't break

bobeal replied:
Yes, I gonna add something to deal with this case.

DavidLeoni commented on pom.xml:
```diff
@@ -243,6 +243,12 @@
 			<version>4.3.1</version>
+
+		<!-- to auto-detect mime type of uploaded resources -->
+		<dependency>
```
Little note: since tika is only used internally it would be better to shade it, see maven-shade-plugin section of the pom. But this is something I can do when I'm going to release the thing

bobeal replied:
As I'm not familiar with Maven, I would appreciate if you could add it :-)

bobeal committed:
Fix missing end tag in pom.xml

bobeal committed:
Add a test case for large files

bobeal committed:
Fix handling of optional / required fields when uploading a resource file

bobeal committed:
Enhancements to uploaded file management

- Handle passing a null file to the setUpload method
- Set mime type and format guessing as an option
- Throw an exception if mime type and format guessing failed

bobeal commented:
Hello,

Here are some more commits to fix what we talked about.

Regards,

Benoit.

bobeal commented:
Hi,

Any news on this PR ? Do you plan to merge it soon ?

Regards,

Benoit.

DavidLeoni commented:
Sorry for the delay, I got busy with deadlines. 

Thank you very much for the pull request, it's a very nice improvement. I only added some documentation and `CkanResourceBase.setUpload(File)`  to still have a bean-style setter.
